### PR TITLE
Pass labels rather than OCM component / artifact

### PR DESCRIPTION
**Pull request by zkdev**
Labels are used to find issues, thus updating, deleting, and so on. As preparation for cfg-element policy violation issues, this hard contract based on OCM has to be resovled as cfg-elements do not belong to an OCM component. Instead, pass down labels directly.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: github/compliance/issue.py [hunk 1]
```diff
@@ -85,23 +90,13 @@ def repository_labels(
 
 @github.retry.retry_and_throttle
 def enumerate_issues(
-    component: cm.Component | None,
-    artifact: cm.Artifact | None,
+    labels: frozenset[str],
```
Comment (ccwienk) [inline github/compliance/issue.py:93]: hmm, actually I do not think we need a frozenset - any iterable should do

File: github/compliance/issue.py [hunk 2]
```diff
@@ -85,23 +90,13 @@ def repository_labels(
...
     known_issues: typing.Sequence[github3.issues.issue.ShortIssue],
-    issue_type: str | None,
```
Comment (ccwienk) [inline github/compliance/issue.py:91]: why remove the issue_type?

File: github/compliance/issue.py [hunk 3]
```diff
@@ -66,7 +71,7 @@ def repository_labels(
     artifact: cm.Artifact | None,
     issue_type: str | None=_label_bdba,
     extra_labels: typing.Iterable[str]=None
-):
+) -> typing.Generator[str, None, None]:
```
Comment (ccwienk) [inline github/compliance/issue.py:74]: is `repository_labels` still used?

File: github/compliance/issue.py [hunk 4]
```diff
@@ -115,54 +110,38 @@ def filter_relevant_issues(issue):
...
-    artifact: cm.Artifact,
+    labels: frozenset[str],
     repository: github3.repos.Repository,
-    body:str,
-    title:typing.Optional[str],
+    body: str,
+    title: str,
     assignees: typing.Iterable[str]=(),
     milestone: github3.issues.milestone.Milestone=None,
-    issue_type: str=_label_bdba,
```
Comment (ccwienk) [inline github/compliance/issue.py:125]: ditto

File: github/compliance/issue.py [hunk 5]
```diff
@@ -115,54 +110,38 @@ def filter_relevant_issues(issue):
...
 def _update_issue(
-    component: cm.Component,
-    artifact: cm.Artifact,
+    labels: frozenset[str],
     body:str,
     title:typing.Optional[str],
     issue: github3.issues.Issue,
     milestone: github3.issues.milestone.Milestone=None,
     assignees: typing.Iterable[str]=(),
-    issue_type: str=_label_bdba,
```
Comment (ccwienk) [inline github/compliance/issue.py:164]: ditto

File: github/compliance/issue.py [hunk 6]
```diff
@@ -192,42 +164,39 @@ def _update_issue(
...
+    labels: frozenset[str],
     repository: github3.repos.Repository,
-    body:str,
+    body: str,
     known_issues: typing.Iterable[github3.issues.issue.ShortIssue],
-    title:str=None,
+    title: str,
     assignees: typing.Iterable[str]=(),
     milestone: github3.issues.milestone.Milestone=None,
-    issue_type: str=_label_bdba,
```
Comment (ccwienk) [inline github/compliance/issue.py:203]: ditto

File: github/compliance/issue.py [hunk 7]
```diff
@@ -192,42 +164,39 @@ def _update_issue(
...
     extra_labels: typing.Iterable[str]=None,
```
Comment (ccwienk) [inline github/compliance/issue.py:174]: why did you remove extra_labels from the other functions, but keep it here?

File: github/compliance/issue.py [hunk 8]
```diff
@@ -192,42 +164,39 @@ def _update_issue(
...
-            artifact=artifact,
+            labels=labels,
             known_issues=known_issues,
-            issue_type=issue_type,
             state='open',
         )
     )
     if (issues_count := len(open_issues)) > 1:
-        raise RuntimeError(f'more than one open issue found for {component.name=}{artifact.name=}')
+        raise RuntimeError(f'more than one open issue found for {labels=}')
```
Comment (ccwienk) [inline github/compliance/issue.py:185]: hmm, that's not nice - those labels are quite cumbersome to read, considering the component/artefact-names are only there as hash-digests

Thus, I think it should be better to change the passed-in "target-identifier", like so:
- instead of component and resource, use `cnudie.iter.Node` (or for a first iteration, we might only accept ResourceNode)
- for cfg-elements, we might accept cfg-element

I am considering something like so:

```
def enumerate_issues(
 target: cnudie.iter.ResourceNode | model.base.NamedModelElement
 .. <other parameters, as they were defined before>
):
 labels = labels_from_target(target=target)
```

def labels_from_target(target: cnudie.iter.ResourceNode | model.base.NamedModelElement):
 if isinstance(target, RN):
   <code from before this change>
 elif isinstance(NME):
  <read element's name and typename to create labels>
 else:
  raise NotImplementedError(target)
```

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: github/compliance/issue.py (hunk 2 above)
Comment (zkdev) [reply github/compliance/issue.py:91]: `issue_type` is only used to get repository labels
> yield f'cicd/{issue_type}'

as we pass in labels already, `issue_type` is not required in this func anymore

File: github/compliance/issue.py (hunk 3 above)
Comment (zkdev) [reply github/compliance/issue.py:74]: yes, `repository_labels` yields _all_ labels for a repo, e.g. `cicd/auto-generated` AND the actual artifact digest label.

Comment (zkdev) [reply github/compliance/issue.py:74]: I separated the digest calculation from repo label yielding as a preparation

File: github/compliance/issue.py (hunk 1 above)
Comment (zkdev) [reply github/compliance/issue.py:93]: We at least have to use collections, see line 103

File: github/compliance/issue.py (hunk 7 above)
Comment (zkdev) [reply github/compliance/issue.py:174]: The other func retrieved repo labels based off component+artifact and the passed extra labels.
As we now pass in labels directly, I merged those to only pass one `labels` arg.

This func however has a special semantic for extra labels, that is considering the passed `preserve_labels_regexes`. Therefore I kept the separation here.

File: github/compliance/issue.py (hunk 8 above)
Comment (zkdev) [reply github/compliance/issue.py:185]: We have to at least implement `cnudie.iter.ResourceNode` and `cnudie.iter.SourceNode` targets as checkmarx scans are in place already.

How about introducing a dataclass?

```python
@dc
class Target:
    target: ResourceNode | SourceNode | ...

    # could be used for logging
    @property
    def name(self) -> str:
        if isinstance(cnudie.iter.ResourceNode):
            return f'{self.target.component.name}:{self.target.resource.name}'

        if isinstance(cnudie.iter.SourceNode):
            return f'{self.target.component.name}:{self.target.source.name}'

    # determining target digest label here, so repository_labels could take an optional
    # digest_label callback and yield "common" labels (cicd/auto-generated, ...)
    def digest_label(self) -> str:
        if isinstance(self.target, cnudie.iter.ResourceNode):
            return artifact_digest_label(
                component=self.target.component,
                artifact=self.target.resource,
            )
        if isinstance(self.target, cnudie.iter.SourceNode):
            return artifact_digest_label(
                component=self.target.component,
                artifact=self.target.source,
            )
        # can be extended for efg-element
```

Comment (ccwienk) [reply github/compliance/issue.py:185]: well, let's support ResourceNode and SourceNode in parallel, then - why introduce a new class for model elements that already have one?

Comment (zkdev) [reply github/compliance/issue.py:185]: > why introduce a new class for model elements that already have one?

to group functions for "targets" like `name()` or `digest_label()`.

Comment (ccwienk) [reply github/compliance/issue.py:185]: I consider this avoidable boilerplate.

just define a function that accepts a target and returns labels for the target, and - if required multiple times - a second function returning the display-name.

the structure as suggested above mixes different concerns, for which even dedicated functions already exist.

if at all, in order to reduce the amount of functions that need to dispatch between different target-types (from two to one), one could consider to define a dataclass holding name and labels. tbh I do not see much value here, and such an approach will not lead to a better codebase

**Review by ccwienk — APPROVED** [review]
